File: app-sdk/src/main/java/code/sdk/ui/WebPresenter.java
package code.sdk.ui;


import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.net.Uri;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.text.TextUtils;
import android.webkit.ValueCallback;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;

import java.io.File;

import code.sdk.R;
import code.sdk.bridge.BridgeCallback;
import code.sdk.bridge.JsBridge;
import code.sdk.common.PermissionUtils;
import code.sdk.common.ShareUtil;
import code.sdk.core.Constant;
import code.sdk.core.manager.AdjustManager;
import code.sdk.core.util.CocosPreferenceUtil;
import code.sdk.core.util.DeviceUtil;
import code.sdk.core.util.FileUtil;
import code.sdk.core.util.PreferenceUtil;
import code.sdk.core.util.UIUtil;
import code.sdk.core.util.URLUtilX;
import code.sdk.download.DownloadTask;
import code.sdk.util.PromotionImageSynthesizer;
import code.util.IOUtil;
import code.util.ImageUtil;
import code.util.LogUtil;

public class WebPresenter {
    public static final String TAG = WebPresenter.class.getSimpleName();

    private WebActivity mWebViewActivity;
    private String mUrl = "";
    private boolean isGame = false;
    private String mWebGlScript;
    private final int msg_check_feature = 20001;
    private final int msg_check_webView_compatibility = 20002;
    private String mImageUrl;

    public final String systemWebViewPackage = "com.google.android.webview";
    private final String miniSystemWebViewVersion = "64.0.3282.29";
    private final int miniSystemWebViewVersionCode = 328202950;
    public final String[] PERMISSIONS_STORAGE = {Manifest.permission.WRITE_EXTERNAL_STORAGE};
    private final Handler mHandler = new Handler(Looper.getMainLooper()) {

        @Override
        public void handleMessage(@NonNull Message msg) {
            if (msg.what == msg_check_feature) {
                String gameUrl = (String) msg.obj;
                handleCheckGameFeature(gameUrl);
            } else if (msg.what == msg_check_webView_compatibility) {
                handleCheckWebViewCompatibility();
            }
        }
    };


    private final BridgeCallback mJsBridgeCb = new BridgeCallback() {
        @Override
        public void openUrlByBrowser(String url) {
            try {
                Intent intent = new Intent(Intent.ACTION_VIEW);
                intent.setData(Uri.parse(url));
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                mWebViewActivity.startActivity(intent);
            } catch (Exception e) {
            }
        }

        @Override
        public void openUrlByWebView(String url) {
            try {
                Intent intent = new Intent(mWebViewActivity, WebActivity.class);
                intent.putExtra(WebActivity.KEY_URL, url);
                mWebViewActivity.startActivity(intent);
            } catch (Exception e) {
            }
        }

        @Override
        public void openApp(String target, String fallbackUrl) {
            try {
                mWebViewActivity.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(target)));
            } catch (Exception e) {
                e.printStackTrace();
                openUrlByBrowser(fallbackUrl);
            }
        }

        @Override
        public void loadUrl(String url) {
            UIUtil.runOnUiThread(() -> {
                mWebViewActivity.getWebView().loadUrl(url);
            });
        }

        @Override
        public void goBack() {
            UIUtil.runOnUiThread(() -> {
                if (mWebViewActivity.getWebView().canGoBack()) {
                    mWebViewActivity.getWebView().goBack();
                } else {
                    mWebViewActivity.finish();
                }
            });
        }

        @Override
        public void close() {
            UIUtil.runOnUiThread(() -> mWebViewActivity.finish());
        }

        @Override
        public void refresh() {
            UIUtil.runOnUiThread(() -> mWebViewActivity.getWebView().reload());
        }

        @Override
        public void clearCache() {
            UIUtil.runOnUiThread(() -> mWebViewActivity.getWebView().clearCache(true));
        }

        @Override
        public void saveImage(String url) {
            mImageUrl = url;
            if (PermissionUtils.checkStoragePermissions(mWebViewActivity)) {
                File dir = mWebViewActivity.getExternalFilesDir(Environment.DIRECTORY_DCIM);
                FileUtil.ensureDirectory(dir);
                DownloadTask.getInstance().download(url, dir.getAbsolutePath(), new DownloadTask.OnDownloadListener() {
                    @Override
                    public void onDownloadSuccess(File saveFile) {
                        boolean succeed = saveFile != null && saveFile.length() > 0;
                        if (succeed) {
                            ImageUtil.triggerScanning(saveFile);
                        }
                        LogUtil.d(TAG, "saveImage - download succeed = " + succeed);
                        saveImageDone(succeed);
                    }

                    @Override
                    public void onDownloading(int progress) {
                        LogUtil.d(TAG, "saveImage - downloading = " + progress + "%");
                    }

                    @Override
                    public void onDownloadFailed() {
                        LogUtil.w(TAG, "saveImage - download failed");
                        saveImageDone(false);
                    }
                });
            } else {
                ActivityCompat.requestPermissions(mWebViewActivity, PERMISSIONS_STORAGE, mWebViewActivity.REQUEST_SAVE_IMAGE);
            }
        }

        @Override
        public void saveImageDone(boolean succeed) {
            UIUtil.runOnUiThread(() -> {
                LogUtil.d(TAG, "saveImageDone = " + succeed);
                String script = succeed ? "Listener.send('SAVE_IMAGE_SUCCEED');" : "Listener.send('SAVE_IMAGE_FAILED');";
                mWebViewActivity.webViewEvaluatescript(script);
            });
        }

        @Override
        public void savePromotionMaterialDone(boolean succeed) {
            UIUtil.runOnUiThread(() -> {
                LogUtil.d(TAG, "savePromotionMaterialDone = " + succeed);
                String script = succeed ? "Listener.send('SAVE_PROMOTION_MATERIAL_SUCCEED');" : "Listener.send('SAVE_PROMOTION_MATERIAL_FAILED');";
                mWebViewActivity.webViewEvaluatescript(script);
            });
        }

        @Override
        public void synthesizePromotionImage(String qrCodeUrl, int size, int x, int y) {
            mWebViewActivity.mPromotionQrCodeUrl = qrCodeUrl;
            mWebViewActivity.mPromotionSize = size;
            mWebViewActivity.mPromotionX = x;
            mWebViewActivity.mPromotionY = y;
            if (PermissionUtils.checkStoragePermissions(mWebViewActivity)) {
                new PromotionImageSynthesizer(mWebViewActivity, qrCodeUrl, size, x, y, this).execute();
            } else {
                ActivityCompat.requestPermissions(mWebViewActivity, PERMISSIONS_STORAGE, mWebViewActivity.REQUEST_SYNTHESIZE_PROMOTION_IMAGE);
            }
        }

        @Override
        public void synthesizePromotionImageDone(boolean succeed) {
            UIUtil.runOnUiThread(() -> {
                LogUtil.d(TAG, "synthesizePromotionImageDone = " + succeed);
                String script = succeed ? "Listener.send('SYNTHESIZE_PROMOTION_IMAGE_SUCCEED');" : "Listener.send('SYNTHESIZE_PROMOTION_IMAGE_FAILED');";
                mWebViewActivity.webViewEvaluatescript(script);
            });
        }

        @Override
        public void onHttpDnsHttpResponse(String requestId, String response) {
            UIUtil.runOnUiThread(() -> {
                LogUtil.d(TAG, "[HttpDns] onHttpDnsHttpResponse: %s", response);
                String script = String.format("Listener.send('HTTPDNS_HTTP_RESPONSE', '%s');", response);
                mWebViewActivity.webViewEvaluatescript(script);
            });
        }

        @Override
        public void onHttpDnsWsResponse(String requestId, String response) {
            UIUtil.runOnUiThread(() -> {
                LogUtil.d(TAG, "[HttpDns] onHttpDnsWsResponse: %s", response);
                String script = String.format("Listener.send('HTTPDNS_WS_RESPONSE', '%s');", response);
                mWebViewActivity.webViewEvaluatescript(script);
            });
        }

        @Override
        public void shareUrl(String url) {
            ShareUtil.sendText(mWebViewActivity, url);
        }

        @Override
        public void loginFacebook() {
        }

        @Override
        public void logoutFacebook() {
        }

        @Override
        public void preloadPromotionImageDone(boolean succeed) {
            UIUtil.runOnUiThread(() -> {
                LogUtil.d(TAG, "preloadPromotionImageDone = " + succeed);
                String script = succeed ? "Listener.send('PRELOAD_PROMOTION_IMAGE_SUCCEED');" : "Listener.send('PRELOAD_PROMOTION_IMAGE_FAILED');";
                mWebViewActivity.webViewEvaluatescript(script);
            });
        }

        @Override
        public void shareToWhatsApp(String text, File file) {
            ShareUtil.shareToWhatsApp(mWebViewActivity, text, file);
        }
    };

    public WebPresenter(WebActivity activity) {
        this.mWebViewActivity = activity;
    }

    public void init(boolean isGame, String url) {
        this.isGame = isGame;
        this.mUrl = url;
        mWebGlScript = IOUtil.readRawContent(mWebViewActivity.getBaseContext(), R.raw.webgl_script);
        LogUtil.d(TAG, "WebGlScript=" + mWebGlScript);
        JsBridge jsBridge = new JsBridge(mJsBridgeCb);
        mWebViewActivity.getWebView().addJavascriptInterface(jsBridge, JsBridge.getJsBridgeName());

        Uri uri = Uri.parse(mUrl);
        // hover menu
        mWebViewActivity.setShowHoverMenu(uri.getBooleanQueryParameter(Constant.QUERY_PARAM_HOVER_MENU, false));
        // nav bar
        mWebViewActivity.setShowNavBar(uri.getBooleanQueryParameter(Constant.QUERY_PARAM_NAV_BAR, false));
        // save cutout
        mWebViewActivity.setSafeCutout(uri.getBooleanQueryParameter(Constant.QUERY_PARAM_SAFE_CUTOUT, false));
        // screen orientation
        String screenOrientation = uri.getQueryParameter(Constant.QUERY_PARAM_ORIENTATION);
        if (Constant.PORTRAIT.equals(screenOrientation) || Constant.UNSPECIFIED.equals(screenOrientation)) {
            mWebViewActivity.setScreenOrientation(screenOrientation);
        }

        if (isGame) {
            AdjustManager.trackEventAccess(null);
        }
    }

    /**
     * 检查游戏特征值包括(延迟1min检查)
     * _int_chn
     * _int_brand_code
     */
    public void checkGameFeature(String gameUrl) {
        if (isGame) {
            mHandler.removeMessages(msg_check_feature);
            Message msg = Message.obtain();
            msg.what = msg_check_feature;
            msg.obj = gameUrl;
            mHandler.sendMessageDelayed(msg, 60000);
        }
    }

    public void clearGameFeature() {
        if (isGame) {
            CocosPreferenceUtil.putString(CocosPreferenceUtil.KEY_INT_CHN, "");
            CocosPreferenceUtil.putString(CocosPreferenceUtil.KEY_INT_BRAND_CODE, "");
        }
    }

    private void handleCheckGameFeature(String gameUrl) {
        String chn = CocosPreferenceUtil.getString(CocosPreferenceUtil.KEY_INT_CHN);
        String brandCode = CocosPreferenceUtil.getString(CocosPreferenceUtil.KEY_INT_BRAND_CODE);
        LogUtil.d(TAG, "check game feature: %s=%s, %s=%s", CocosPreferenceUtil.KEY_INT_CHN, chn, CocosPreferenceUtil.KEY_INT_BRAND_CODE, brandCode);
        if (TextUtils.isEmpty(chn) && TextUtils.isEmpty(brandCode)) {//特征值为空，不符合游戏特征(判定为不是我们自己的游戏)
            //当前加载的url跟缓存的url一样，并且不符合游戏特征，则清除掉缓存url
            String cacheGameUrl = PreferenceUtil.readGameUrl();
            LogUtil.d(TAG, "check game feature: is not validate game page: %s", gameUrl);
            if (isSameBaseUrl(gameUrl, cacheGameUrl)) {
                LogUtil.d(TAG, "check game feature: clear cached url: %s", cacheGameUrl);
                PreferenceUtil.saveGameUrl(null);
            } else {
                LogUtil.d(TAG, "check game feature: abort clearing cached url: %s", cacheGameUrl);
            }
        } else {
            LogUtil.d(TAG, "check game feature: is validate game page: %s", gameUrl);
        }
    }

    /* Check WebView Compatibility START */

    /**
     * 检查WebView兼容性（两个方面：WebGl和Android System WebView版本）
     */
    public void checkWebViewCompatibility() {
        if (isGame) {
            mHandler.removeMessages(msg_check_webView_compatibility);
            mHandler.sendEmptyMessageDelayed(msg_check_webView_compatibility, 5000);
        }
    }

    public void jsBrSaveImage() {
        mJsBridgeCb.saveImage(mImageUrl);
    }

    public void jsBrSaveImageDone(Boolean succeed) {
        mJsBridgeCb.saveImageDone(succeed);
    }

    public void jsBrSynthesizePromotionImageDone(boolean succeed) {
        mJsBridgeCb.synthesizePromotionImageDone(succeed);
    }

    public BridgeCallback getJsBridge() {
        return mJsBridgeCb;
    }

    private void handleCheckWebViewCompatibility() {
        if (TextUtils.isEmpty(mWebGlScript)) {
            LogUtil.d(TAG, "abort checking WebView compatibility, WebGlScript is empty");
            return;
        }
        LogUtil.d(TAG, "start checking WebView compatibility...");
        boolean isWebViewCompatible = isWebViewCompatible();
        mWebViewActivity.getWebView().evaluateJavascript(mWebGlScript, new ValueCallback<String>() {
            @Override
            public void onReceiveValue(String value) {//js与native交互的回调函数
                LogUtil.d(TAG, "isWebGlEnable = " + value);
                if ("false".equals(value) || !isWebViewCompatible) {
                    UIUtil.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            mWebViewActivity.showWebViewUpdateDialog();
                        }
                    });
                }
            }
        });
    }

    /**
     * 游戏需要的Android System WebView最小版本是：64.0.3282.29 - 328202950
     * 小于该版本，内核会报以下错误：
     * Uncaught SyntaxError: Invalid regular expression: /(?<bundle>.+)UiLanguage/:
     *
     * @return
     */
    private boolean isWebViewCompatible() {
        PackageInfo packageInfo = DeviceUtil.getPackageInfo(mWebViewActivity.getBaseContext(), systemWebViewPackage);
        boolean compatible = false;
        if (packageInfo != null) {
            if (packageInfo.versionCode >= miniSystemWebViewVersionCode) {
                compatible = true;
                LogUtil.i(TAG, "[Android System WebView] version: %s(%d) compatible for cocos game",
                        packageInfo.versionName, packageInfo.versionCode);
            } else {
                LogUtil.e(TAG, "[Android System WebView] version: %s(%d) not compatible for cocos game, need upgrade to %s(%d) or higher",
                        packageInfo.versionName, packageInfo.versionCode,
                        miniSystemWebViewVersion,
                        miniSystemWebViewVersionCode);
            }
        } else {
            LogUtil.i(TAG, "[Android System WebView] not installed");
        }
        return compatible;
    }

    private boolean isSameBaseUrl(String url1, String url2) {
        if (TextUtils.isEmpty(url1) || TextUtils.isEmpty(url2)) {
            return false;
        }
        String baseUrl1 = URLUtilX.getBaseUrl(url1);
        String baseUrl2 = URLUtilX.getBaseUrl(url2);
        if (!baseUrl1.endsWith("/")) {
            baseUrl1 = baseUrl1 + "/";
        }
        if (!baseUrl2.endsWith("/")) {
            baseUrl2 = baseUrl2 + "/";
        }
        return baseUrl1.equals(baseUrl2);
    }

    public void onDestroy() {
        mHandler.removeMessages(msg_check_webView_compatibility);
        mHandler.removeMessages(msg_check_feature);
    }
}

File: app-sdk/src/main/java/code/sdk/ui/WebPresenter.kt
package code.sdk.ui

import android.Manifest
import android.content.Intent
import android.net.Uri
import android.os.Environment
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.text.TextUtils
import android.webkit.ValueCallback
import androidx.core.app.ActivityCompat
import code.sdk.R
import code.sdk.bridge.BridgeCallback
import code.sdk.bridge.JsBridge
import code.sdk.bridge.JsBridgeCore
import code.sdk.bridge.JsBridgeCore.Companion.getJsBridgeName
import code.sdk.common.PermissionUtils.checkStoragePermissions
import code.sdk.common.ShareUtil.sendText
import code.sdk.common.ShareUtil.shareToWhatsApp
import code.sdk.core.Constant
import code.sdk.core.manager.AdjustManager
import code.sdk.core.util.CocosPreferenceUtil
import code.sdk.core.util.DeviceUtil
import code.sdk.core.util.FileUtil
import code.sdk.core.util.PreferenceUtil
import code.sdk.core.util.UIUtil
import code.sdk.core.util.URLUtilX
import code.sdk.download.DownloadTask
import code.sdk.download.DownloadTask.OnDownloadListener
import code.sdk.ui.WebActivity
import code.sdk.util.PromotionImageSynthesizer
import code.util.IOUtil.readRawContent
import code.util.ImageUtil.triggerScanning
import code.util.LogUtil.d
import code.util.LogUtil.e
import code.util.LogUtil.i
import code.util.LogUtil.w
import java.io.File

class WebPresenter(private val mWebViewActivity: WebActivity) {
   private val TAG = WebPresenter::class.java.simpleName

    val PERMISSIONS_STORAGE = arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE)
    val systemWebViewPackage = "com.google.android.webview"
    private var mUrl = ""
    private var isGame = false
    private var mWebGlScript: String? = null
    private val msg_check_feature = 20001
    private val msg_check_webView_compatibility = 20002
    private var mImageUrl: String? = null
    private val miniSystemWebViewVersion = "64.0.3282.29"
    private val miniSystemWebViewVersionCode = 328202950
    private val mHandler: Handler = object : Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            if (msg.what == msg_check_feature) {
                val gameUrl = msg.obj as String
                handleCheckGameFeature(gameUrl)
            } else if (msg.what == msg_check_webView_compatibility) {
                handleCheckWebViewCompatibility()
            }
        }
    }
    val jsBridge = object : BridgeCallback {
        override fun openUrlByBrowser(url: String?) {
            try {
                val intent = Intent(Intent.ACTION_VIEW)
                intent.setData(Uri.parse(url))
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                mWebViewActivity.startActivity(intent)
            } catch (_: Exception) {
            }
        }

        override fun openUrlByWebView(url: String?) {
            try {
                val intent = Intent(mWebViewActivity, WebActivity::class.java)
                intent.putExtra(WebActivity.KEY_URL, url)
                mWebViewActivity.startActivity(intent)
            } catch (_: Exception) {
            }
        }

        override fun openDataByWebView(
            type: Int,
            orientation: String?,
            hover: Boolean,
            data: String?
        ) {
            try {
                val intent = Intent(mWebViewActivity, WebActivity::class.java)
                intent.putExtra(WebActivity.KEY_TYPE, type)
                intent.putExtra(WebActivity.KEY_ORIENTATION, orientation)
                intent.putExtra(WebActivity.KEY_HOVER, hover)
                intent.putExtra(WebActivity.KEY_URL, data)
                mWebViewActivity.startActivity(intent)
            } catch (_: Exception) {
            }
        }

        override fun openApp(target: String?, fallbackUrl: String?) {
            try {
                mWebViewActivity.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(target)))
            } catch (e: Exception) {
                e.printStackTrace()
                openUrlByBrowser(fallbackUrl)
            }
        }

        override fun loadUrl(url: String?) {
            if (!url.isNullOrEmpty())
                UIUtil.runOnUiThread { mWebViewActivity.webView.loadUrl(url) }
        }

        override fun goBack() {
            UIUtil.runOnUiThread {
                if (mWebViewActivity.webView.canGoBack()) {
                    mWebViewActivity.webView.goBack()
                } else {
                    mWebViewActivity.finish()
                }
            }
        }

        override fun close() {
            UIUtil.runOnUiThread { mWebViewActivity.finish() }
        }

        override fun refresh() {
            UIUtil.runOnUiThread { mWebViewActivity.webView.reload() }
        }

        override fun clearCache() {
            UIUtil.runOnUiThread { mWebViewActivity.webView.clearCache(true) }
        }

        override fun saveImage(url: String?) {
            mImageUrl = url
            if (checkStoragePermissions(mWebViewActivity)) {
                val dir = mWebViewActivity.getExternalFilesDir(Environment.DIRECTORY_DCIM)
                FileUtil.ensureDirectory(dir)
                DownloadTask.mInstance.download(url, dir!!.absolutePath, object : OnDownloadListener {
                    override fun onDownloadSuccess(saveFile: File) {
                        val succeed = saveFile.length() > 0
                        if (succeed) {
                            triggerScanning(saveFile)
                        }
                        d(TAG, "saveImage - download succeed = $succeed")
                        saveImageDone(succeed)
                    }

                    override fun onDownloading(progress: Int) {
                        d(TAG, "saveImage - downloading = $progress%")
                    }

                    override fun onDownloadFailed() {
                        w(TAG, "saveImage - download failed")
                        saveImageDone(false)
                    }
                })
            } else {
                ActivityCompat.requestPermissions(
                    mWebViewActivity,
                    PERMISSIONS_STORAGE,
                    mWebViewActivity.REQUEST_SAVE_IMAGE
                )
            }
        }

        override fun saveImageDone(succeed: Boolean) {
            UIUtil.runOnUiThread {
                d(TAG, "saveImageDone = $succeed")
                val script =
                    if (succeed) "Listener.send('SAVE_IMAGE_SUCCEED');" else "Listener.send('SAVE_IMAGE_FAILED');"
                mWebViewActivity.webViewEvaluatescript(script)
            }
        }

        override fun savePromotionMaterialDone(succeed: Boolean) {
            UIUtil.runOnUiThread {
                d(TAG, "savePromotionMaterialDone = $succeed")
                val script =
                    if (succeed) "Listener.send('SAVE_PROMOTION_MATERIAL_SUCCEED');" else "Listener.send('SAVE_PROMOTION_MATERIAL_FAILED');"
                mWebViewActivity.webViewEvaluatescript(script)
            }
        }

        override fun synthesizePromotionImage(qrCodeUrl: String?, size: Int, x: Int, y: Int) {
            mWebViewActivity.mPromotionQrCodeUrl = qrCodeUrl
            mWebViewActivity.mPromotionSize = size
            mWebViewActivity.mPromotionX = x
            mWebViewActivity.mPromotionY = y
            if (checkStoragePermissions(mWebViewActivity)) {
                PromotionImageSynthesizer(mWebViewActivity, qrCodeUrl, size, x, y, this).execute()
            } else {
                ActivityCompat.requestPermissions(
                    mWebViewActivity,
                    PERMISSIONS_STORAGE,
                    mWebViewActivity.REQUEST_SYNTHESIZE_PROMOTION_IMAGE
                )
            }
        }

        override fun synthesizePromotionImageDone(succeed: Boolean) {
            UIUtil.runOnUiThread {
                d(TAG, "synthesizePromotionImageDone = $succeed")
                val script =
                    if (succeed) "Listener.send('SYNTHESIZE_PROMOTION_IMAGE_SUCCEED');" else "Listener.send('SYNTHESIZE_PROMOTION_IMAGE_FAILED');"
                mWebViewActivity.webViewEvaluatescript(script)
            }
        }

        override fun onHttpDnsHttpResponse(requestId: String?, response: String?) {
            UIUtil.runOnUiThread {
                d(TAG, "[HttpDns] onHttpDnsHttpResponse: %s", response)
                val script =
                    String.format("Listener.send('HTTPDNS_HTTP_RESPONSE', '%s');", response)
                mWebViewActivity.webViewEvaluatescript(script)
            }
        }

        override fun onHttpDnsWsResponse(requestId: String?, response: String?) {
            UIUtil.runOnUiThread {
                d(TAG, "[HttpDns] onHttpDnsWsResponse: %s", response)
                val script = String.format("Listener.send('HTTPDNS_WS_RESPONSE', '%s');", response)
                mWebViewActivity.webViewEvaluatescript(script)
            }
        }

        override fun shareUrl(url: String?) {
            sendText(mWebViewActivity, url)
        }

        override fun loginFacebook() {}
        override fun logoutFacebook() {}
        override fun preloadPromotionImageDone(succeed: Boolean) {
            UIUtil.runOnUiThread {
                d(TAG, "preloadPromotionImageDone = $succeed")
                val script =
                    if (succeed) "Listener.send('PRELOAD_PROMOTION_IMAGE_SUCCEED');" else "Listener.send('PRELOAD_PROMOTION_IMAGE_FAILED');"
                mWebViewActivity.webViewEvaluatescript(script)
            }
        }

        override fun shareToWhatsApp(text: String?, file: File?) {
            shareToWhatsApp(mWebViewActivity, text, file)
        }
    }

    fun init(isGame: Boolean, url: String) {
        this.isGame = isGame
        mUrl = url
        mWebGlScript = readRawContent(mWebViewActivity.baseContext, R.raw.webgl_script)
        d(TAG, "WebGlScript=$mWebGlScript")
        val jsBridge = JsBridge(jsBridge)
        mWebViewActivity.webView.addJavascriptInterface(jsBridge, JsBridgeCore.getJsBridgeName())
        val uri = Uri.parse(mUrl)
        // hover menu
        mWebViewActivity.setShowHoverMenu(
            uri.getBooleanQueryParameter(
                Constant.QUERY_PARAM_HOVER_MENU,
                false
            )
        )
        // nav bar
        mWebViewActivity.setShowNavBar(
            uri.getBooleanQueryParameter(
                Constant.QUERY_PARAM_NAV_BAR,
                false
            )
        )
        // save cutout
        mWebViewActivity.setSafeCutout(
            uri.getBooleanQueryParameter(
                Constant.QUERY_PARAM_SAFE_CUTOUT,
                false
            )
        )
        // screen orientation
        val screenOrientation = uri.getQueryParameter(Constant.QUERY_PARAM_ORIENTATION)
        if (Constant.PORTRAIT == screenOrientation || Constant.UNSPECIFIED == screenOrientation) {
            mWebViewActivity.setScreenOrientation(screenOrientation)
        }
        if (isGame) {
            AdjustManager.trackEventAccess(null)
        }
    }

    /**
     * 检查游戏特征值包括(延迟1min检查)
     * _int_chn
     * _int_brand_code
     */
    fun checkGameFeature(gameUrl: String?) {
        if (isGame) {
            mHandler.removeMessages(msg_check_feature)
            val msg = Message.obtain()
            msg.what = msg_check_feature
            msg.obj = gameUrl
            mHandler.sendMessageDelayed(msg, 60000)
        }
    }

    fun clearGameFeature() {
        if (isGame) {
            CocosPreferenceUtil.putString(CocosPreferenceUtil.KEY_INT_CHN, "")
            CocosPreferenceUtil.putString(CocosPreferenceUtil.KEY_INT_BRAND_CODE, "")
        }
    }

    private fun handleCheckGameFeature(gameUrl: String) {
        val chn = CocosPreferenceUtil.getString(CocosPreferenceUtil.KEY_INT_CHN)
        val brandCode = CocosPreferenceUtil.getString(CocosPreferenceUtil.KEY_INT_BRAND_CODE)
        d(
            TAG,
            "check game feature: %s=%s, %s=%s",
            CocosPreferenceUtil.KEY_INT_CHN,
            chn,
            CocosPreferenceUtil.KEY_INT_BRAND_CODE,
            brandCode
        )
        if (TextUtils.isEmpty(chn) && TextUtils.isEmpty(brandCode)) { //特征值为空，不符合游戏特征(判定为不是我们自己的游戏)
            //当前加载的url跟缓存的url一样，并且不符合游戏特征，则清除掉缓存url
            val cacheGameUrl = PreferenceUtil.readGameUrl()
            d(TAG, "check game feature: is not validate game page: %s", gameUrl)
            if (isSameBaseUrl(gameUrl, cacheGameUrl)) {
                d(TAG, "check game feature: clear cached url: %s", cacheGameUrl)
                //PreferenceUtil.saveGameUrl(null);
            } else {
                d(TAG, "check game feature: abort clearing cached url: %s", cacheGameUrl)
            }
        } else {
            d(TAG, "check game feature: is validate game page: %s", gameUrl)
        }
    }
    /* Check WebView Compatibility START */
    /**
     * 检查WebView兼容性（两个方面：WebGl和Android System WebView版本）
     */
    fun checkWebViewCompatibility() {
        if (isGame) {
            mHandler.removeMessages(msg_check_webView_compatibility)
            mHandler.sendEmptyMessageDelayed(msg_check_webView_compatibility, 5000)
        }
    }

    fun jsBrSaveImage() {
        jsBridge.saveImage(mImageUrl)
    }

    fun jsBrSaveImageDone(succeed: Boolean?) {
        jsBridge.saveImageDone(succeed!!)
    }

    fun jsBrSynthesizePromotionImageDone(succeed: Boolean) {
        jsBridge.synthesizePromotionImageDone(succeed)
    }

    private fun handleCheckWebViewCompatibility() {
        if (TextUtils.isEmpty(mWebGlScript)) {
            d(TAG, "abort checking WebView compatibility, WebGlScript is empty")
            return
        }
        d(TAG, "start checking WebView compatibility...")
        val isWebViewCompatible = isWebViewCompatible
        mWebViewActivity.webView
            .evaluateJavascript(mWebGlScript!!) { value -> //js与native交互的回调函数
                d(TAG, "isWebGlEnable = $value")
                if ("false" == value || !isWebViewCompatible) {
                    UIUtil.runOnUiThread { mWebViewActivity.showWebViewUpdateDialog() }
                }
            }
    }

    /**
     * 游戏需要的Android System WebView最小版本是：64.0.3282.29 - 328202950
     * 小于该版本，内核会报以下错误：
     * Uncaught SyntaxError: Invalid regular expression: /(?<bundle>.+)UiLanguage/:
     *
     * @return
    </bundle> */
    private val isWebViewCompatible: Boolean
         get() {
            val packageInfo =
                DeviceUtil.getPackageInfo(mWebViewActivity.baseContext, systemWebViewPackage)
            var compatible = false
            if (packageInfo != null) {
                if (packageInfo.versionCode >= miniSystemWebViewVersionCode) {
                    compatible = true
                    i(
                        TAG, "[Android System WebView] version: %s(%d) compatible for cocos game",
                        packageInfo.versionName, packageInfo.versionCode
                    )
                } else {
                    e(
                        TAG,
                        "[Android System WebView] version: %s(%d) not compatible for cocos game, need upgrade to %s(%d) or higher",
                        packageInfo.versionName,
                        packageInfo.versionCode,
                        miniSystemWebViewVersion,
                        miniSystemWebViewVersionCode
                    )
                }
            } else {
                i(TAG, "[Android System WebView] not installed")
            }
            return compatible
        }

    private fun isSameBaseUrl(url1: String, url2: String): Boolean {
        if (TextUtils.isEmpty(url1) || TextUtils.isEmpty(url2)) {
            return false
        }
        var baseUrl1 = URLUtilX.getBaseUrl(url1)
        var baseUrl2 = URLUtilX.getBaseUrl(url2)
        if (!baseUrl1.endsWith("/")) {
            baseUrl1 = "$baseUrl1/"
        }
        if (!baseUrl2.endsWith("/")) {
            baseUrl2 = "$baseUrl2/"
        }
        return baseUrl1 == baseUrl2
    }

    fun onDestroy() {
        mHandler.removeMessages(msg_check_webView_compatibility)
        mHandler.removeMessages(msg_check_feature)
    }
}
